structuredheadings/plugin.js

```diff
@@ -222,24 +183,162 @@
       setupStyles(editor);
 
-      //Dialogs
-      //eslint-disable-next-line new-cap
-      editor.addCommand("selectStyle", new CKEDITOR.dialogCommand("selectStyle", {
-        startDisabled: true,
-        contextSensitive: true,
-        refresh: function () {
-          if (isEmpty(editor.config.autonumberStyles)) {
-            this.setState(CKEDITOR.TRISTATE_DISABLED);
+      //Format Dropdown
+      editor.ui.addRichCombo("NumFormats", {
+        label: "Formats",
+        title: "Numbering Formats",
+        toolbar: "styles,7",
+        allowedContent: editor.config.numberedElements.concat(["pre", "p", "div"]),
+
+        panel: {
+          css: [ CKEDITOR.skin.getPath("editor") ].concat(editor.config.contentsCss),
+          multiSelect: false,
+          attributes: { "aria-label": "Numbering Formats" }
+        },
+
+        init: function () {
+
+          this.startGroup("Formats");
+
+          this.add("p",
+            elementStyles.p.buildPreview(
+              "Normal Text"
+            ),
+            "Normal Text"
+          );
+          for (var key in editor.config.numberedElements) {
+            this.add(editor.config.numberedElements[key],
+              elementStyles[ editor.config.numberedElements[key] ].buildPreview(
+                "Header " + editor.config.numberedElements[key].slice(1)
+              ),
+              "Header " + editor.config.numberedElements[key].slice(1));
+          }
+          this.add("pre",
+              elementStyles.pre.buildPreview(
+                  "Formatted Text"
+                ),
+            "Preformatted Text"
+          );
+        },
+
+        onClick: function (value) {
+          editor.applyStyle(elementStyles[ value ]);
+          setLevel(editor, editor.elementPath().block);
+          editor.execCommand("reapplyStyle");
+          if (value === "p") {
+            this.setValue(value, "Normal Text");
+          } else if (value === "pre") {
+            this.setValue(value, "Formatted Text");
           } else {
-            this.setState(CKEDITOR.TRISTATE_OFF);
+            this.setValue(value, "Header " + editor.config.numberedElements[
+                editor.config.numberedElements.indexOf(value)
+              ].slice(1));
           }
+
+        },
+
+        onRender: function () {
+          editor.on("selectionChange", function (ev) {
+            var currentTag = this.getValue();
+            var elementPath = ev.data.path;
+            var elementList = editor.config.numberedElements.concat(["pre", "p", "div"]);
+            for (var tag in elementList) {
+              if (elementStyles[elementList[tag]].checkActive(elementPath, editor)) {
+                if (elementList[tag] !== currentTag) {
+                  if (elementList[tag] === "p") {
+                    this.setValue(elementList[tag], "Normal Text");
+                  } else if (elementList[tag] === "pre") {
+                    this.setValue(elementList[tag], "Formatted Text");
```

Comment:
elementList is referencing config.numberedElements which needs to be a flat array for purposes of integer indexes. In this case these tags are being added with .concat within this function and aren't a part of the original list. Setting up an object would be taking place still within this function a couple lines up and I don't think would be a benefit.